server/datastore/mysql/hosts.go

```diff
@@ -1399,6 +1427,220 @@ func (ds *Datastore) SetOrUpdateMunkiVersion(ctx context.Context, hostID uint, v
 	)
 }
 
+func (ds *Datastore) replaceHostMunkiIssues(ctx context.Context, hostID uint, msgToID map[[2]string]uint) error {
+	// This needs an efficient way to check if the batch of messages are the same
+	// as existing ones, as this should be a common case (i.e. Munki does not run
+	// *that* often, so it is likely that the host reports the same messages for
+	// some time, or none at all). The approach is as follows:
+	//
+	// * Read a COUNT of new IDs (those to be saved) and a COUNT of old IDs
+	//   (those to be removed) from the read replica.
+	// * If COUNT(new) == len(newIDs) and COUNT(old) == 0, no write is required.
+	// * If COUNT(old) > 0, delete those obsolete ids.
+	// * If COUNT(new) < len(newIDs), insert those missing ids.
+	//
+	// In the best scenario, a single statement is executed on the read replica,
+	// and in the worst case, 3 statements are executed, 2 on the primary.
+	//
+	// Of course this is racy, as the check is done in the replica and the write
+	// is not transactional, but this is not an issue here as host-reported data
+	// is eventually consistent in nature and that data is reported at regular
+	// intervals.
```

Comment:
Mmmman, this feels way too complex for what it's doing in the end. I wonder if they can be sorted alphabetically and then hashed. Stored as one row... but it also feels like I thought about this for 10 mins vs how much time you invested in this... so, take this comment with a huge package of salt.

Reply:
There's probably a bit of misunderstanding about what this is and how we use it - we can't just store a hash of messages that a host has, we need to know exactly what messages are associated with a host, compute aggregated stats for them, and list all hosts that have a specific message ID. This is a one-to-many relationship between a host and munki issues.

This is similar to software/host_software (which is arguably more complex in implementation, and it potentially hits the primary with 4 statements for each host), with the difference that the data is big (long-ish strings) and some extra care were taken to prevent hitting the primary too hard. As mentioned in the comment, I optimized for what should be the more common case that the list hasn't changed in-between two calls for a host, by doing a quick count check on the replica.

I honestly don't feel like this is overly complicated, and it did quite well performance-wise in load testing which was a concern, but of course I'm happy to consider any improvements in clarity or even a different approach that maintains the requirements mentioned above.

Reply:
Certainly. It's one of those problems that sounds simple when you think about it the amount of time that I have so far, and then as you dive in it gets more tricky. And that depth changes how the solution feels. Like you said, the software one is a similar example in that sense, I've hit myself against that particular wall more than once because my brain tells me "how can this not be 1 simple sql line and move on??"

Reply:
Yeah for sure :D Also, whenever we ingest host-related data, we have to be concerned about performance at scale, so there's that too (vs e.g. delete all for that host and re-insert all on each call, which sure would be simpler, but would likely cause issues on the primary at scale).